### ISSUE TITLE
C# compiler should not report CS8780 on receiving variable in switch on union w/ generic type

### ISSUE THREAD

**Issue by Hafthor** [dotnet/roslyn#85270]
**Version Used**: 11.0.100-rc.1.26425.128

**Steps to Reproduce**:

1. Create C# 15 project using .NET 11 RC1
2. Add the code
```csharp
public union Result<T>(T, Exception) {
  public string Stringize() => this switch {
    T theResult => theResult.ToString(),
    Exception ex => ex.Message
  };
}
```

**Diagnostic Id**: CS8780: A variable may not be declared within a 'not' or an 'or' pattern or a union matching involving matching against either the instance, or its underlying value.

**Expected Behavior**: No error

**Actual Behavior**: CS8780 on receiving 'theResult' variable (after generic T). Did not occur on August .NET 11 Preview.

**Comment by RikkiGibson** [comment]
This looks like an issue that was fixed by #84897, this should make it into next release candidate.

**Comment by Hafthor** [comment]
Thanks @RikkiGibson. Should we add a test for `switch { T t`? I didn't see anything like that in UnionsTests.cs.

**Comment by Hafthor** [comment]
I did confirm that this is fixed in the current Roslyn code. Thanks!
```csharp
[Fact]
public void UnionType_00()
{
    var src = @"union U<T>(T) { T M() => this switch { T t => t }; }";
    var comp = CreateCompilation([src, UnionAttributeSource, IUnionSource]);
    comp.VerifyEmitDiagnostics();
}
```
